Fix a crash when 'read -u' is given an invalid fd

JohnoKing opened the pull request:
File descriptors that are too far out of range cause `read -u` to crash. The following example will generate two crashes:

```sh
$ ksh -c 'read -u2000000' || ksh -c 'read -u-2000000'
```

The fix is to error out when the given file descriptor is out of range. This bugfix is from Tomas Klacko, although it was
modified to use `sh_iovalidfd` and reject numbers greater than `INT_MAX`:
https://www.mail-archive.com/ast-developers@lists.research.att.com/msg01912.html
The question about `shp->fdstatus[-1]` only applies to ksh93v- (ksh93u+ doesn't have any references to `shp->fdstatus[-1]` afaict).

McDutchie commented on src/cmd/ksh93/bltins/read.c:
```diff
@@ -127,7 +127,11 @@ int	b_read(int argc,char *argv[], Shbltin_t *context)
+		if(opt_info.num<0 || opt_info.num>INT_MAX)
+			errormsg(SH_DICT,ERROR_exit(1),e_file,opt_info.arg); /* reject negative and out of range numbers */
 		fd = (int)opt_info.num;
+		if(fd>=shp->gd->lim.open_max && !sh_iovalidfd(shp,fd))
```
Is there any reason not to test for both conditions here, and save compiling in an `errormsg` call? Like:

```c
	if(fd < 0 || fd > INT_MAX || (fd>=shp->gd->lim.open_max && !sh_iovalidfd(shp,fd)))
```

JohnoKing replied:
The pull request has been updated with the optimization, although the first two tests must compare against `opt_info.num` as `fd` is only an `int`, which is too small for handling numbers outside of the 32-bit limit.

McDutchie replied:
Of course, good point, thanks.

JohnoKing committed:
Fix a crash when 'read -u' is given an invalid fd

File descriptors that are too far out of range will cause the
read builtin to crash. The following example will generate
two crashes:

$ ksh -c 'read -u 2000000' || ksh -c 'read -u-2000000'

The fix is to error out when the given file descriptor is out
of range. This bugfix is from Tomas Klacko, although it was
modified to use 'sh_iovalidfd' and reject numbers greater
than 'INT_MAX':
https://www.mail-archive.com/ast-developers@lists.research.att.com/msg01912.html
The question about 'shp->fdstatus[-1]' only applies to ksh93v-
(ksh93u+ doesn't have any references to 'shp->fdstatus[-1]').

src/cmd/ksh93/bltins/read.c:
- File descriptors that are out of range should be rejected
  with an error message (like invalid file descriptors that
  are in range). The seemingly redundant check for negative
  numbers is there because out of range negative numbers also
  cause memory faults despite the later 'fd<0' check.

src/cmd/ksh93/tests/io.sh:
- Add three tests for attempting 'read -u' on various invalid
  file descriptor numbers.
